NEXT-31522 - Added documentation for entity attributes (#1390)
shopware/docs: guides/plugins/plugins/framework/data-handling/entities-via-attributes.md

---
nav:
  title: Entities via attributes
  position: 1

---

# Entities via attributes

> Since shopware v6.6.3.0 it is possible to register entities via php attributes. This is a new way to register entities in Shopware 6. This guide will show you how to do this.

Reviewer: ```suggestion
Since Shopware v6.6.3.0 it is possible to register entities via PHP attributes. This guide will show you how to do this.
```

Removed redundant text

## Define the entity

First, you need to define your entity. This is done by creating a new class and adding the `Entity` attribute to it. The attribute requires the `name` parameter, which is the name of the entity. The `name` parameter is required and must be unique.

> Also, you have to extend the `EntityStruct` class and define a primary key. The primary key is defined by adding the `PrimaryKey` attribute to a property. In theory, the primary key can be of any type, but it is recommended to use a `UUID`. 

Reviewer 2: `EntityStruct` is an alias of `Shopware\Core\Framework\DataAbstractionLayer\Entity`. IMO, it would make sense to de-alias it and use alias for `Shopware\Core\Framework\DataAbstractionLayer\Attribute\Entity as EntityAttribute` instead?

> First, you need to define your entity. This is done by creating a new class and adding the `Entity` attribute to it. The attribute requires the `name` parameter, which is the name of the entity. The `name` parameter is required and must be unique.
> 
> Also, you have to extend the `EntityStruct` class and define a primary key. The primary key is defined by adding the `PrimaryKey` attribute to a property. In theory, the primary key can be of any type, but it is recommended to use a `UUID`. 

Reviewer: ```suggestion
First, you need to define your entity. This is done by creating a new class extending `EntityStruct` and adding the `Entity` attribute to it. The attribute requires the `name` parameter, which is the name of the entity. The `name` parameter is required and must be unique.

You have to define a primary key. The primary key is defined by adding the `PrimaryKey` attribute to a property. In theory, the primary key can be of any type, but it is recommended to use a `UUID`. 
```

Linguistically makes more sense

```php
<?php

> namespace Shopware\Tests\Integration\Core\Framework\DataAbstractionLayer\fixture;

Reviewer 3: maybe we should some Foo/Bar namespace name in the docs, instead of the "real" namespace of our tests? same in the services.xml

Reviewer 2: Maybe use some expected namespace for Plugins?

use Shopware\Core\Framework\DataAbstractionLayer\Entity as EntityStruct;
use Shopware\Core\Framework\DataAbstractionLayer\Attribute\Entity;

#[Entity('attribute_entity')]
class AttributeEntity extends EntityStruct
{
    #[PrimaryKey]
    #[Field(type: FieldType::UUID)]
    public string $id;
}
```

This is the most basic entity definition. You can add more properties and attributes to the entity. For example, you can add the `Field` attribute to a property to define the type of the property.

## Register the entity

To register the entity, you have to add this class to the DI container in the `services.xml` file. This is done by adding the `shopware.entity` tag to the service definition. 

```xml
<service id="Shopware\Tests\Integration\Core\Framework\DataAbstractionLayer\fixture\AttributeEntity">
    <tag name="shopware.entity"/>
</service>
```

> That's it. You entity is registered and you can read and write data to it over the DAL. Over the tag, shopware automatically registers an `EntityDefinition` for the entity and an `EntityRepository`.

Reviewer: ```suggestion
That's it. You entity is registered and you can read and write data to it over the DAL. Using the tag, Shopware automatically registers an `EntityDefinition` and `EntityRepository` for the entity.
```

## Field Types

> To define more fields, you typically use the `Field` attribute. The `Field` attribute requires the `type` parameter, which is the type of the field. The type can be any of the `FieldType` constants. 

Reviewer 3: "typically" ? is there another way? :grin: 

To define more fields, add a new public property and add the `Field` attribute to it.

Author: yes here are other ways, other field classes.

```php

#[Entity('attribute_entity')]
class AttributeEntity extends EntityStruct
{
    #[PrimaryKey]
    #[Field(type: FieldType::UUID)]
    public string $id;
    
     #[Field(type: FieldType::STRING)]
    public string $string;

    #[Field(type: FieldType::TEXT)]
    public ?string $text = null;

    #[Field(type: FieldType::INT)]
    public ?int $int;

    // ...
}
```

All field types are defined in the [`FieldType`](https://github.com/shopware/shopware/blob/trunk/src/Core/Framework/DataAbstractionLayer/Attribute/FieldType.php) class.

> We also provide a list of special field-types, which implement a specific behavior. They have an own php attribute class, for example, the `AutoIncrement` or `ForeignKey` field. 

Reviewer: ```suggestion
We also provide a list of special field types, which implement a specific behavior. They have their own PHP attribute class, for example the `AutoIncrement` or `ForeignKey` field. 
```

```php
#[Entity('attribute_entity')]
class AttributeEntity extends EntityStruct
{
    #[PrimaryKey]
    #[Field(type: FieldType::UUID)]
    public string $id;
    
    #[AutoIncrement]
    public int $autoIncrement;

>     #[ForeignKey(entity: 'currency')]

Reviewer 3: isn't it possible to use constants here? :thinking: 

```suggestion
    #[ForeignKey(entity: \Shopware\Core\System\Currency\CurrencyDefinition::ENTITY_NAME)]
```

Author: yes it's possible, but it makes it harder to read. I would not recommend this. we have a validator and if you failed in the name, it throws an exception.

    public ?string $foreignKey;
}
```

## JSON fields

> If you want to store json data in a field, with its own validation and serialization logic, you can use the `Serialized` attribute, and define its own serializer class:

Reviewer 4: ```suggestion
If you want to store JSON data in a field with its own validation and serialization logic, you can use the `Serialized` attribute and define its own serializer class:
```

> ## Json fields 
> 
> If you want to store json data in a field, with an own validation and serialize logic, you can use the `Serialized` attribute, and define an own serializer class:

Reviewer: JSON is an abbreviation and the standard way to write it is upprecase.

```suggestion
## JSON fields

If you want to store json data in a field, with its own validation and serialization logic, you can use the `Serialized` attribute, and define its own serializer class:
```

```php
#[Entity('attribute_entity')]
class AttributeEntity extends EntityStruct
{
    #[PrimaryKey]
    #[Field(type: FieldType::UUID)]
    public string $id;
    
    #[Serialized(serializer: PriceFieldSerializer::class)]
    public ?PriceCollection $serialized = null;
}
```

## Custom Fields

If you want to allow custom fields, you can use the `CustomField` attribute. 

```php
<?php

#[Entity('attribute_entity')]
class AttributeEntity extends EntityStruct
{
    #[PrimaryKey]
    #[Field(type: FieldType::UUID)]
    public string $id;
    
    /**
     * @var array<string, mixed>
     */
    #[CustomFields]
    public array $customFields;
}

```

> ## Api encoding

Reviewer: ```suggestion
## API encoding
```

> Each field of an entity is by default not exposed in the API. To expose a field in the API, you have to set the `api` parameter of the `Field` attribute to `true`, or to one of the scopes you want to allow. 

Reviewer: ```suggestion
By default, each field of an entity is not exposed in the API. To expose a field in the API, you must set the api parameter of the Field attribute to true or specify one of the scopes you want to allow.
```

```php
<?php

use Shopware\Core\Framework\Api\Context\AdminApiSource;
use Shopware\Core\Framework\Api\Context\SalesChannelApiSource;

#[Entity('attribute_entity')]
class AttributeEntity extends EntityStruct
{
    #[PrimaryKey]
    #[Field(type: FieldType::UUID, api: true)]
    public string $id;

    #[Field(type: FieldType::STRING)]
    public string $notExposed;

    #[Field(type: FieldType::STRING, api: true)]
    public string $everywhere;

    #[Field(type: FieldType::STRING, api: [AdminApiSource::class])]
    public string $adminOnly;

    #[Field(type: FieldType::STRING, api: [SalesChannelApiSource::class])]
    public string $storeOnly;
```

## Translated fields

> To also support shopware translations for your entity, you can set the `translated` property of the `Field` attribute to `true`. This will automatically create a `TranslatedField` for the field and
> registers a `EntityTranslationDefinition` for you.
> 
> You can also define a `Translations` attribute on a property to allow loading all translations of the entity.

Reviewer: ```suggestion
To support Shopware translations for your entity, set the `translated` property of the `Field` attribute to `true`. This will automatically create a `TranslatedField` for the field and
register an `EntityTranslationDefinition` for you.

Additionally, you can define a `Translations` attribute on a property to enable loading all translations of the entity.
```

Notice: Properties with the `translated` flag must be nullable. 

```php
#[Entity('attribute_entity')]
class AttributeEntity extends EntityStruct
{
    #[PrimaryKey]
    #[Field(type: FieldType::UUID)]
    public string $id;

    #[Field(type: FieldType::STRING, translated: true)]
    public ?string $string = null;

    /**
     * @var array<string, ArrayEntity>
     */
    #[Translations]
    public array $translations;
}
```

## Required fields

> Each field which is not type hinted as nullable is required by default. But you can also mark a field as required by adding the `Required` attribute to the field. This will automatically add a validation rule to the field.
> This is necessary for fields which are marked as `translated`, because the translated fields need to be nullable.

Reviewer: ```suggestion
By default, any field that is not type-hinted as nullable is required. However, you can explicitly mark a field as required by adding the `Required` attribute to it. This will automatically add a validation rule to the field.
This is necessary for fields marked as `translated`, as translated fields must be nullable.
```

```php
#[Entity('attribute_entity')]
class AttributeEntity extends EntityStruct
{
    #[PrimaryKey]
    #[Field(type: FieldType::UUID)]
    public string $id;

    #[Required]
    #[Field(type: FieldType::STRING, translated: true)]
    public ?string $required = null;
}
```

## Associations

> It is also possible to define associations between entities. You can you one of the following four association types `OneToOne`, `OneToMany`, `ManyToOne` and `ManyToMany`.    

Reviewer: ```suggestion
It is also possible to define associations between entities. You can use one of the following four association types: `OneToOne`, `OneToMany`, `ManyToOne` or `ManyToMany`.    
```

    #[ForeignKey(entity: 'currency')]
    public ?string $currencyId = null;

    #[ForeignKey(entity: 'currency')]
    public ?string $followId = null;

    #[ManyToOne(entity: 'currency')]
    public ?CurrencyEntity $currency = null;

    #[OneToOne(entity: 'currency')]
    public ?CurrencyEntity $follow = null;

    /**
     * @var array<string, AttributeEntityAgg>
     */
    #[OneToMany(entity: 'attribute_entity_agg', ref: 'attribute_entity_id')]
    public ?array $aggs = null;

    /**
     * @var array<string, CurrencyEntity>
     */
    #[ManyToMany(entity: 'currency')]
    public ?array $currencies = null;
}
```

All the associations are defined as a nullable array property. The key of the array is the id of the associated entity. The value is the associated entity itself.

You can also typehint to many associations with the `EntityCollection` class.

## Getter & Setter, Translations and Collections

With this new pattern, we removed the need for getter and setter methods. The properties are public and can be accessed directly.
Also, you don't have to define any `EntityTranslationDefinition` or `EntityCollection` anymore, which reduces the boilerplate code. 

> 

Reviewer: remove empty line
